zebra/rt_netlink.c

```diff
@@ -1584,6 +1607,239 @@ static bool _netlink_route_build_singlepath(const struct prefix *p,
 		}
 	}
 
+	if (fpm && nexthop->nh_srv6) {
+		if (nexthop->nh_srv6->seg6local_action !=
+		    ZEBRA_SEG6_LOCAL_ACTION_UNSPEC) {
+			struct zebra_srv6 *srv6 = zebra_srv6_get_default();
+			struct zebra_vrf *zvrf;
+			bool locator_found = false;
+			struct srv6_locator *locator;
+			struct listnode *node;
+			struct rtattr *nest, *inner_nest;
+			const struct seg6local_context *ctx;
+
+			ctx = &nexthop->nh_srv6->seg6local_ctx;
+
+			if (!nl_attr_put16(nlmsg, req_size, RTA_ENCAP_TYPE,
+					   FPM_NH_ENCAP_SRV6_LOCAL_SID))
```

Comment:
so the design here is "pick a new encap number, and use that to contain the various tlvs with srv6 info", is that right?
is that a robust choice - will the OS-based netlink messaging never support this information? the original goal of the netlink encoded FPM messaging was to be able to use a single encoding or marshalling code-base consistently, but this sort of pushes that aside and introduces a different approach.

Reply:
@mjstapp Thanks for your review. 

Yes, we pick a new encap number and we use that to contain the TLVs with SRv6 information.

I know `dplane_fpm_nl` is designed to share zebra's Netlink encoding functions but unfortunately we can't reuse the existing encoding functions in this case.

The reason is that the SRv6 kernel's data model is very different from the SRv6 SONiC's data model.

---

Let's make an example.

When we encapsulate a packet using SRv6, we add an outer IPv6 header and we need to choose an IPv6 address to use as source address in the outer IPv6 header.

This is implemented differently in the Linux kernel and in SONiC.
In the Linux kernel, we have a source address configured globally in the system and we use that address for all the SRv6 encapsulated packets. For this reason, the Netlink message used to install a route in the kernel does not contain the source address.

Instead, in SONiC each SRv6 route has its own IPv6 source address. Therefore, the Netlink message used to install an SRv6 route in SONiC must contain the source address.